agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,345 @@
+     */
+    log2_num_buckets = 8;
+  } else if (log2_num_buckets > 24) {
+    /*
+     * Basic sanity check: it's extremely unlikely that we'll ever need a
+     * hashmap for the user function wraprecs that has more than 2^24 buckets.
+     */
+    log2_num_buckets = 24;
+  }
+
+  hashmap
+      = (nr_php_wraprec_hashmap_t*)nr_malloc(sizeof(nr_php_wraprec_hashmap_t));
+  hashmap->dtor_func = dtor_fn;
+  hashmap->log2_num_buckets = log2_num_buckets;
+  hashmap->buckets = (nr_wraprecs_bucket_t**)nr_calloc(
+      (1 << log2_num_buckets), sizeof(nr_wraprecs_bucket_t*));
+  hashmap->elements = 0;
+
+  return hashmap;
+}
+
+nr_php_wraprec_hashmap_t* nr_php_wraprec_hashmap_create_buckets(
+    size_t buckets,
+    nr_php_wraprec_hashmap_dtor_fn_t dtor_fn) {
+  size_t actual_buckets;
+
+  if (0 == buckets) {
+    actual_buckets = 0;
+  } else {
+    /*
+     * Calculate ceil(log2(buckets)), since that's what
+     * nr_hashmap_create_internal requires.
+     */
+
+    actual_buckets = 1;
+    while (((size_t)(1 << actual_buckets)) < buckets) {
+      actual_buckets += 1;
+    }
+  }
+
+  return nr_php_wraprec_hashmap_create_internal(actual_buckets, dtor_fn);
+}
+
+static void nr_destroy_wraprecs_bucket(
+    nr_wraprecs_bucket_t** bucket_ptr,
+    nr_php_wraprec_hashmap_dtor_fn_t dtor_func) {
+  nr_wraprecs_bucket_t* bucket = *bucket_ptr;
+
+  if (dtor_func) {
+    (dtor_func)(bucket->wraprec);
+  }
+  nr_realfree((void**)bucket_ptr);
+}
+
+nr_php_wraprec_hashmap_stats_t nr_php_wraprec_hashmap_destroy(
+    nr_php_wraprec_hashmap_t** hashmap_ptr) {
+  nr_php_wraprec_hashmap_stats_t stats = {};
+  size_t count;
+  nr_php_wraprec_hashmap_t* hashmap;
+  size_t i;
+  size_t bucket_items_cnt;
+
+  if ((NULL == hashmap_ptr) || (NULL == *hashmap_ptr)) {
+    return stats;
+  }
+  hashmap = *hashmap_ptr;
+
+  stats.elements = hashmap->elements;
+  stats.collisions_min = stats.elements;
+
+  count = nr_count_buckets(hashmap);
+  for (i = 0; i < count; i++) {
+    nr_wraprecs_bucket_t* bucket = hashmap->buckets[i];
+    bucket_items_cnt = 0;
+
+    if (bucket) {
+      stats.buckets_used++;
+    }
+
+    while (bucket) {
+      nr_wraprecs_bucket_t* next = bucket->next;
+
+      bucket_items_cnt++;
+
+      nr_destroy_wraprecs_bucket(&bucket, hashmap->dtor_func);
+      bucket = next;
+    }
+
+    if (0 != bucket_items_cnt && bucket_items_cnt < stats.collisions_min) {
+      stats.collisions_min = bucket_items_cnt;
+    }
+    if (bucket_items_cnt > stats.collisions_max) {
+      stats.collisions_max = bucket_items_cnt;
+    }
+    if (bucket_items_cnt > 1) {
+      stats.buckets_with_collisions++;
+      stats.collisions_mean += bucket_items_cnt;
+    }
+  }
+
+  if (0 != stats.buckets_with_collisions) {
+    stats.collisions_mean /= stats.buckets_with_collisions;
+  }
+
+  nr_free(hashmap->buckets);
+  nr_realfree((void**)hashmap_ptr);
+
+  return stats;
+}
+
+static bool nr_zf_is_unnamed_closure(const zend_function* zf) {
```

Comment:
Could consider making this inline.

Reply:
Good idea! Done in 3930b15d3c4e8552676dea1b87bdf3c5a3b6f33f.